widget_weather.go

```diff
@@ -0,0 +1,229 @@
+		return "", nil
+	}
+
+	wttr := strings.Split(w.response, " ")
+	if len(wttr) != 2 {
+		return "", fmt.Errorf("can't parse weather response: %s", w.response)
+	}
+
+	return strings.Replace(wttr[1], "+", "", 1), nil
+}
+
+// Ready returns true when weather data is available.
+func (w *WeatherData) Ready() bool {
+	w.responseMutex.RLock()
+	defer w.responseMutex.RUnlock()
+
+	return len(w.response) > 0
+}
+
+// Fetch retrieves weather data when required.
+func (w *WeatherData) Fetch() {
+	w.responseMutex.Lock()
+	defer w.responseMutex.Unlock()
+
+	if time.Since(w.refresh) < time.Minute*15 {
+		return
+	}
+	// fmt.Println("Refreshing weather data...")
+
+	url := "http://wttr.in/" + w.location + "?format=%x+%t" + formatUnit(w.unit)
+
+	resp, err := http.Get(url) //nolint:gosec
+	if err != nil {
+		fmt.Println("can't fetch weather data:", err)
+		return
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Println("can't read weather data:", err)
+		return
+	}
+
+	w.refresh = time.Now()
+	w.response = string(body)
+}
+
+// NewWeatherWidget returns a new WeatherWidget.
+func NewWeatherWidget(bw BaseWidget, opts WidgetConfig) *WeatherWidget {
+	bw.setInterval(opts.Interval, 60000)
+
+	var location, unit, theme string
+	_ = ConfigValue(opts.Config["location"], &location)
+	_ = ConfigValue(opts.Config["unit"], &unit)
+	_ = ConfigValue(opts.Config["theme"], &theme)
+	var color color.Color
+	_ = ConfigValue(opts.Config["color"], &color)
+	var flatten bool
+	_ = ConfigValue(opts.Config["flatten"], &flatten)
+
+	if color == nil {
+		color = DefaultColor
+	}
+
+	if theme == "" {
+		theme = "openmoji"
+	}
+
+	return &WeatherWidget{
+		BaseWidget: bw,
+		data: WeatherData{
+			location: location,
+			unit:     unit,
+		},
+		color:   color,
+		theme:   theme,
+		flatten: flatten,
+	}
+}
+
+// Update renders the widget.
+func (w *WeatherWidget) Update(dev *streamdeck.Device) error {
+	go w.data.Fetch()
+	if !w.data.Ready() {
+		return nil
+	}
+
+	cond, err := w.data.Condition()
+	if err != nil {
+		return err
+	}
+	temp, err := w.data.Temperature()
+	if err != nil {
+		return err
+	}
+
+	var iconName string
+	switch cond {
+	case "mm", "mmm": // cloudy
+		iconName = "cloudy"
+	case "m": // partly cloudy
+		iconName = "partly_cloudy"
+	case "=": // fog
+		iconName = "fog"
+	case "///", "//", "x", "x/": // rain
+		iconName = "rain"
+	case "/", ".": // light rain
+		iconName = "light_rain"
+	case "**", "*/*": // heavy snow
+		iconName = "heavy_snow"
+	case "*", "*/": // light snow
+		iconName = "light_snow"
+	case "/!/": // thunder
+		iconName = "thunder"
+	case "!/", "*!*": // thunder rain
+		iconName = "thunder_rain"
+	case "o": // sunny
+		if time.Now().Hour() < 7 || time.Now().Hour() > 21 {
+			iconName = "moon"
+		} else {
+			iconName = "sun"
+		}
+	default:
+		return w.render(dev, nil)
+	}
+
+	weatherIcon := weatherImage("assets/weather/" + w.theme + "/" + iconName + ".png")
```

Comment:
Instead of manually concatenating a path here, we should use `filepath.Join("assets", "weather", w.theme, iconName+".png")`.